<?xml version="1.0"?>
<launch>
  <!-- These three arguments come from the CLI -->
  <arg name="simulate"/>
  <arg name="map_name"/>
  <arg name="track_name"/>
  
  <group ns="controllers">
    <include file="$(find pathing)/launch/controllers.launch">
      <arg name="simulate" value="$(arg simulate)"/>
    </include>
  </group>

  <!-- Will launch a simulated map and lidar if the CLI requests it -->
  <group ns="simulation">
    <!-- Launch a map from the maps folder as instructed from the command line -->
    <arg name="map" default="$(find pathing)/maps/$(arg map_name)/$(arg map_name).yaml"/>
    <arg name="track" default="$(find pathing)/tracks/$(arg track_name)/$(arg track_name)_map.yaml"/>
    
    <node pkg="map_server" name="map_server" type="map_server" args="$(arg map)" if="$(eval arg('map_name') != 'no')">
      <remap from="/simulation/map" to="/map"/>
    </node>
    <node pkg="map_server" name="map_server" type="map_server" args="$(arg track)" if="$(eval arg('track_name') != 'no' and arg('map_name') == 'no')">
      <remap from="/simulation/map" to="/map"/>
    </node>

    <!-- Simulate only the lidar -->
    <node pkg="pathing" name="simulator" type="sim_lidar" output="screen" if="$(eval arg('simulate') == 'map')" >
      <rosparam command="load" file="$(find pathing)/params.yaml"/>
    </node>
    
    <!-- Simulate everything-->
    <node pkg="pathing" name="simulator" type="sim_car" output="screen" if="$(eval arg('simulate') == 'all')" >
      <rosparam command="load" file="$(find pathing)/params.yaml"/>
    </node>
    
    <!-- Launch the racecar model -->
    <include file="$(find pathing)/launch/racecar_model.launch"/>
  </group>

  <group ns="pathing">
    <include file="$(find pathing)/launch/pathing.launch"/>
  </group>

  <group ns="algorithms">
    <include file="$(find pathing)/launch/algorithms.launch"/>
  </group>
  
  <group ns="led_control">
    <include file="$(find pathing)/launch/led.launch"/>
  </group>
    
  <!-- Launch node to convert rviz user nav goal to pose with covariance (where covariance is set to 0) -->
  <node pkg="pathing" name="pose_to_posewithcovariance" type="pose_to_posewithcovariance" output="screen">
    <rosparam command="load" file="$(find pathing)/params.yaml"/>
  </node>
  
  <!-- Launch node to flip the scan info 180 degrees about the z axis for the wall follow, gap follow, etc. nodes -->
  <node pkg="pathing" name="lidar_flip" type="lidar_fix" output="screen">
    <rosparam command="load" file="$(find pathing)/params.yaml"/>
  </node>
</launch>
